<a href="https://colab.research.google.com/github/DaudAhmad0303/LeetCode-Problems-Solution-colab/blob/master/LeetCode_Problems_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Random Practise Codes**

In [ ]:
nums = [4,1,2,3,8]
print(nums[::2])
print(nums[::-2])
print(nums[::-3])

[4, 2, 8]
[8, 2, 4]
[8, 1]


# **Random Questions Solved**

## **2164. Sort Even and Odd Indices Independently**

In [ ]:
class Solution:
    def sortEvenOdd(self, nums):
        if len(nums) <= 2:
            return nums
        odd = []
        even = []
        l = len(nums)
        for i in range(len(nums)):
            if i % 2 == 0:
                even.append(nums[i])
            else:
                odd.append(nums[i])
        odd.sort(reverse = True)        
        even.sort()
        nums = []
        for i in range(len(odd)):
            nums.append(even[i])
            nums.append(odd[i])
        if l % 2 != 0:
            nums.append(even[-1])
        return nums
obj = Solution()
print(obj.sortEvenOdd(nums = [4,1,2,3]))

[2, 3, 4, 1]


In [ ]:
class Solution:
    def sortEvenOdd(self, nums):
        odds = nums[::2]
        evens = nums[1::2]
        odds.sort()
        evens.sort()
        evens.reverse()
        odds.append(-1)
        evens.append(-1)
        
        res = []
        for a,b in zip(odds, evens):
            if a >= 0:
                res.append(a)
            if b >= 0:
                res.append(b)
        
        return res
obj = Solution()
print(obj.sortEvenOdd(nums = [4,1,2,3]))

[2, 3, 4, 1]


## **125. Valid Palindrome**

In [ ]:
class Solution:
    def isPalindrome(self, s: str) -> bool:
        st = ""
        for i in range(len(s)):
            if s[i].isalpha() or s[i].isdigit():
                st += s[i].lower()
        print(st)
        return st == st[::-1]
obj = Solution()
print(obj.isPalindrome(s = "A man, a plan, a canal: Panama"))

amanaplanacanalpanama
True


## **Count Operations to Obtain Zero**

In [ ]:
class Solution:
    def countOperations(self, num1: int, num2: int) -> int:
        if num1 != num2:
            return min(num1, num2) + abs(num1-num2)
        elif num1 == 0 or num2 == 0:
            return 0
        else:
            return 1
obj = Solution()
print(obj.countOperations(num1 = 2, num2 = 3))

3


In [ ]:
class Solution:
    def countOperations(self, num1: int, num2: int) -> int:
        op = 0
        while True:
            if num1 > num2:
                num1 -= num2
                op += 1
            elif num2 > num1:
                num2 -= num1
                op += 1
            else:
                num1 = 0
                op += 1
            if num1 == 0 or num2 == 0:
                break
        return op
            


obj = Solution()
print(obj.countOperations(num1 = 79, num2 = 68))

14


## **118. Pascal's Triangle**

In [ ]:
class Solution:
    def generate(self, numRows: int):
        if numRows == 1:
            return [[1]]
        lst = [[1]]
        i = 0
        for c in range(numRows-1):
            lst1 = [1]
            for j in range(len(lst[i])-1):
                lst1.append(lst[i][j] + lst[i][j+1])
            lst.append(lst1 + [1])
            i += 1
        return lst
obj = Solution()
print(obj.generate(30))

[[1], [1, 1], [1, 2, 1], [1, 3, 3, 1], [1, 4, 6, 4, 1], [1, 5, 10, 10, 5, 1], [1, 6, 15, 20, 15, 6, 1], [1, 7, 21, 35, 35, 21, 7, 1], [1, 8, 28, 56, 70, 56, 28, 8, 1], [1, 9, 36, 84, 126, 126, 84, 36, 9, 1], [1, 10, 45, 120, 210, 252, 210, 120, 45, 10, 1], [1, 11, 55, 165, 330, 462, 462, 330, 165, 55, 11, 1], [1, 12, 66, 220, 495, 792, 924, 792, 495, 220, 66, 12, 1], [1, 13, 78, 286, 715, 1287, 1716, 1716, 1287, 715, 286, 78, 13, 1], [1, 14, 91, 364, 1001, 2002, 3003, 3432, 3003, 2002, 1001, 364, 91, 14, 1], [1, 15, 105, 455, 1365, 3003, 5005, 6435, 6435, 5005, 3003, 1365, 455, 105, 15, 1], [1, 16, 120, 560, 1820, 4368, 8008, 11440, 12870, 11440, 8008, 4368, 1820, 560, 120, 16, 1], [1, 17, 136, 680, 2380, 6188, 12376, 19448, 24310, 24310, 19448, 12376, 6188, 2380, 680, 136, 17, 1], [1, 18, 153, 816, 3060, 8568, 18564, 31824, 43758, 48620, 43758, 31824, 18564, 8568, 3060, 816, 153, 18, 1], [1, 19, 171, 969, 3876, 11628, 27132, 50388, 75582, 92378, 92378, 75582, 50388, 27132, 11628, 3876

In [ ]:
class Solution:
    def generate(self, numRows: int):
        if numRows == 1:
            return [[1]]
        elif numRows == 2:
            return [[1],[1,1]]
        lst = [[1],[1,1]]
        i = 1
        for c in range(numRows-2):
            lst1 = [1]
            for j in range(len(lst[i])-1):
                lst1.append(lst[i][j] + lst[i][j+1])
            lst.append(lst1 + [1])
            i += 1
        return lst
obj = Solution()
print(obj.generate(5))

[[1], [1, 1], [1, 2, 1], [1, 3, 3, 1], [1, 4, 6, 4, 1]]


## **551. Student Attendance Record I**

In [ ]:
%%time
class Solution:
    def checkRecord(self, s: str) -> bool:
        if 'LLL' in s:
            return False
        if 'A' in s:
            i = s.index('A')
            s = s[:i] + s[i+1:]
        if 'A' in s:
            return False
        return True
obj = Solution()
print(obj.checkRecord(s = "PPALLP"))

True
CPU times: user 142 µs, sys: 0 ns, total: 142 µs
Wall time: 148 µs


In [ ]:
%%time
class Solution:
    def checkRecord(self, s: str) -> bool:
        if s.count('A') < 2:
            if 'LLL' in s:
                return False
            else:
                return True
        else:
            return False
obj = Solution()
print(obj.checkRecord(s = "PPALLP"))

True
CPU times: user 159 µs, sys: 0 ns, total: 159 µs
Wall time: 125 µs


# **Daily Coding Challenge Februray**

## ***Feb 6:* 80. Remove Duplicates from Sorted Array II**

In [ ]:
class Solution:
    def removeDuplicates(self, nums) -> int:
        l = len(nums)
        prev = nums[0]
        count = 0
        d_count = 0
        for i in range(l-1):
            if d_count >= 2:
                nums.remove(nums[i])
                nums.append(0)
            else:
                count += 1
            if nums[i+1] != nums[i]:
                d_count = 0
            else:
                d_count +=1
        print(nums)
        return 5

In [ ]:
class Solution:
    def removeDuplicates(self, nums) -> int:
        if len(nums) < 3: 
            return len(nums)
        pos = 1
        for i in range(1, len(nums)-1):
            if nums[i-1] != nums[i+1]:
                nums[pos] = nums[i]
                pos += 1
        nums[pos] = nums[-1]
        print(nums)
        return pos + 1

obj = Solution()
print(obj.removeDuplicates(nums = [0,0,1,1,1,1,2,3,3]))

[0, 0, 1, 1, 2, 3, 3, 3, 3]
7


## ***Feb 7:* 389. Find the Difference**

In [ ]:
class Solution:

    def findTheDifference(self, s: str, t: str) -> str:

        sum1 = 0

        sum2 = 0

        for i in s:

            sum1 += ord(i)

        for i in t:

            sum2 += ord(i)

        return chr(sum2 - sum1)

obj = Solution()
print(obj.findTheDifference(s = "abcd", t = "abcde"))

e


## ***Feb 9:* 258. Add Digits**

In [ ]:
def make_list(n):
    lst = list(str(n))
    lst = list(map(int, lst))
    return 
make_list(399)

<class 'int'>


In [ ]:
class Solution:
    def addDigits(self, num: int) -> int:
        def make_list(n):
            lst = list(str(n))
            lst = list(map(int, lst))
            return lst
        while num>9:
            lst = make_list(num)
            val = 0
            for i in lst:
                val += i
            num = val
        return num
obj = Solution()
print(obj.addDigits(398))

2


## ***Feb 9:* 532. K-diff Pairs in an Array**

In [ ]:
class Solution:
    def findPairs(self, nums, k: int) -> int:
        # return the number of unique pairs
        answ=0
        # create a dictionary: d[x]==nums.count(x)
        d={}
        for x in nums:
            if x in d: d[x]+=1
            else:      d[x]=1
        
        if k: # k>0
            answ=sum(x+k in d for x in d.keys())
        else: # k==0
            answ=sum(k>1 for k in d.values())
                
        return answ
obj = Solution()
print(obj.findPairs(nums = [3,1,4,1,5], k = 2))

2


## ***Feb 10:* 560. Subarray Sum Equals K**

In [ ]:
class Solution:
    def subarraySum(self, nums, k: int) -> int:
        d ={0:1}
    
        sums = 0

        count= 0

        for i in range(len(nums)):
            sums+= nums[i]

            if sums-k in d:
                count+= d[sums-k]

            if sums in d:
                d[sums]+=1

            else:
                d[sums] = 1

        return count
obj = Solution()
print(obj.subarraySum(nums = [1,2,3], k = 3))

2


## ***Feb 11:* 567. Permutation in String**

In [ ]:
import itertools
s = 'ABC'
nums = list(s)
# nums = ['ABC']
permutations = list(itertools.permutations(nums))
lst = [''.join(permutation) for permutation in permutations]
print(lst)

['ABC', 'ACB', 'BAC', 'BCA', 'CAB', 'CBA']


In [ ]:
from collections import Counter
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        # Edge Condition
        if len(s1) > len(s2):
            return False
        n1 = len(s1)
        
		# Hashmap for s1
        c1 = Counter(s1)
		
        for i in range(len(s2)-n1+1):
            if Counter(s2[i:i+n1]) == c1:
                       return True
        return False
obj = Solution()
print(obj.checkInclusion(s1 = "ab", s2 = "eidbaooo"))

True


## ***Feb 12:* 127. Word Ladder**

In [ ]:
from collections import *
class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList) -> int:
        if endWord not in wordList or not endWord or not beginWord or not wordList:
            return 0
        L = len(beginWord)
        all_combo_dict = defaultdict(list)
        for word in wordList:
            for i in range(L):
                all_combo_dict[word[:i] + "*" + word[i+1:]].append(word) 
        queue = deque([(beginWord, 1)])
        visited = set()
        visited.add(beginWord)
        while queue:
            current_word, level = queue.popleft()
            for i in range(L):
                intermediate_word = current_word[:i] + "*" + current_word[i+1:]
                for word in all_combo_dict[intermediate_word]:
                    if word == endWord:
                        return level + 1
                    if word not in visited:
                        visited.add(word)
                        queue.append((word, level + 1))
        return 0
obj = Solution()
print(obj.ladderLength(beginWord = "hit", endWord = "cog", wordList = ["hot","dot","dog","lot","log","cog"]))

5


## ***Feb 13:* 78. Subsets**

In [ ]:
class Solution:
    def subsets(self, nums):
        nums.sort()
        lst = [[]]
        for n in nums:
            lst += [i + [n] for i in lst]
        return lst
obj = Solution()
print(obj.subsets(nums = [1,2,3]))

[[], [1], [2], [1, 2], [3], [1, 3], [2, 3], [1, 2, 3]]


## ***Feb 14:* 104. Maximum Depth of Binary Tree**

In [ ]:
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def maxDepth(self, root) -> int:
        def dfs(root, depth):
            if not root: return depth
            return max(dfs(root.left, depth + 1), dfs(root.right, depth + 1))

        return dfs(root, 0)
obj = Solution()

ob2 = TreeNode(9, None, None)
ob4 = TreeNode(7, None, None)
ob5 = TreeNode(15, None, None)
ob3 = TreeNode(20, ob5, ob4)
ob1 = TreeNode(3, ob2, ob3)

print(obj.maxDepth(ob1))

3


## ***Feb 15:*136. Single Number**

In [ ]:
class Solution:
    def singleNumber(self, nums) -> int:
        return (2 * sum(set(nums))) - sum(nums)
obj = Solution()
print(obj.singleNumber(nums = [4,1,2,1,2]))

4


## ***Feb 17:* 39. Combination Sum**

In [ ]:
class Solution:
    def combinationSum(self, candidates, target: int):
        idx_d = {val: idx for idx, val in enumerate(candidates)} # create {num:idx}
        n = len(candidates)
        dp = [[] for _ in range(target+1)]
        for i in range(1, target+1):                         # from 1 to target, O(M)
            for j in range(i):                               # O(M): for all previous calculated numbers
                for comb in dp[j]:                           # O(M) worst: check each of their combinations
                    start_idx = idx_d[comb[-1]]              # use start_idx to avoid repeats
                    for val in candidates[start_idx:]:       # O(N): try all candidates
                        if val + j == i: dp[i].append(comb + [val]) 
            for candidate in candidates:                     # O(N): directly from candidates not from previous result 
                if candidate == i: dp[i].append([candidate])
        return dp[-1]
obj = Solution()
print(obj.combinationSum(candidates = [2,3,6,7], target = 7))

[[2, 2, 3], [7]]


## ***Feb 18:* 402. Remove K Digits**

In [ ]:
class Solution:
    def removeKdigits(self, num: str, k: int) -> str:
        stack = []

        for n in num:
            while stack and k > 0 and stack[-1] > n:
                k -= 1
                stack.pop()
            stack.append(n)

        while stack and k > 0:
            stack.pop()
            k -= 1

        if not stack:
            return "0"

        return str(int("".join(stack)))
obj = Solution()
print(obj.removeKdigits(num = "1432219", k = 3))

1219


## ***Feb 19:* 1675. Minimize Deviation in Array**

In [ ]:
MAX = 10**20
from heapq import *
class Solution:
    def minimumDeviation(self, nums) -> int:
        heap = []
        answer = MAX
        array_length = len(nums)
        
        for num in nums:
            if num % 2 == 0:
                heappush(heap, -num)
            else:
                heappush(heap, -num*2)
        
        _min = -max(heap)

        while True:
            _max = -heappop(heap)
            answer = min(answer, _max-_min)
            
            if _max % 2:
                break

            heappush(heap, -_max//2)
            _min = min(_min, _max//2)

        return answer
obj = Solution()
print(obj.minimumDeviation(nums = [1,2,3,4]))

1


## ***Feb 20:* 1288. Remove Covered Intervals**

In [ ]:
class Solution:
    def removeCoveredIntervals(self, intervals) -> int:
        intervals = sorted(intervals, key = lambda x : (x[0], -x[1]))
        count = len(intervals)
        l = intervals[0][0]
        h = intervals[0][1]
        for i in range(1, len(intervals)):
            if intervals[i][0] < l and intervals[i][1] > h:
                l = intervals[i][0]
                h = intervals[i][1]
                count -= 1
                print(intervals[i], end = ' ')
            elif (intervals[i][0] and intervals[i][1]) in range(l, h+1):
                count -= 1
            elif intervals[i][1] > h:
                h = intervals[i][1]
            elif intervals[i][0] < l:
                l = intervals[i][0]
        return count
                

        
obj = Solution()
intervals = [[1,4],[3,6],[2,8]]
intervals = [[1,4],[2,3]]
intervals = [[34335,39239],[15875,91969],[29673,66453],[53548,69161],[40618,93111]]
# intervals.sort(key=lambda x: [x[0], -x[1]])
# print(intervals)
print(obj.removeCoveredIntervals(intervals))

[40618, 93111]
2


## ***Feb 22:* 171. Excel Sheet Column Number**

## ***Feb 23:* 133. Clone Graph**

## ***Feb 24:* Sort List**

## ***Feb 25:* 165. Compare Version Numbers**

In [ ]:
class Solution:
    def compareVersion(self, version1: str, version2: str) -> int:
        v1 = version1.split(".")
        v2 = version2.split(".")
        
        while v1 or v2:
            v1_val = 0
            v2_val = 0
            if v1: v1_val = int(v1.pop(0))
            if v2: v2_val = int(v2.pop(0))
            
            if v1_val > v2_val: return 1
            if v1_val < v2_val: return -1
            
        return 0
obj = Solution()
print(obj.compareVersion(version1 = "1.01", version2 = "1.001"))

0


## ***Feb 26:* 847. Shortest Path Visiting All Nodes**

In [ ]:
class Solution:
    def shortestPathLength(self, graph) -> int:
        length = len(graph)

        result = 0

        visited_node = []

        queue = []  

        for i in range(length):
          visited_node.append(set([1<<i]))
          queue.append([i,1<<i])

        while queue:

          result = result + 1

          new_queue = []

          for node, value in queue:

            for neigbour_node in graph[node]:

              mid_node = (1<<neigbour_node)|value

              if mid_node not in visited_node[neigbour_node]:

                if mid_node+1 == 1<<length:

                  return result

                visited_node[neigbour_node].add(mid_node)

                new_queue.append([neigbour_node, mid_node])

          queue = new_queue

        return 0
obj = Solution()
print(obj.shortestPathLength(graph = [[1],[0,2,4],[1,3,4],[2],[1,2]]))

4


## ***Feb 27:* 662. Maximum Width of Binary Tree**

In [ ]:
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
from collections import defaultdict
class Solution:
    def widthOfBinaryTree(self, root) -> int:
        di = defaultdict(list)
        def dfs(node, level, column):
            if node:
                di[level].append(column)
                dfs(node.left, level+1, column*2)
                dfs(node.right, level+1, column*2+1)
        dfs(root, 0 , 0)
        return max([max(di[level]) - min(di[level]) +1 for level in di])

root = [1,3,2,5]
o1 = TreeNode()
o2 = TreeNode()
o3 = TreeNode()
o4 = TreeNode()

o1.val = 1
o1.left = o2
o1.right = o3

o2.val = 3
o2.left = o4
o3.val = 2
o4.val = 5
obj = Solution()
print(obj.widthOfBinaryTree(o1))


2


## ***Feb 28:* 228. Summary Ranges**

In [ ]:
class Solution:
    def summaryRanges(self, nums):
        if len(nums) == 0:   return []
        ret = []
        temp = [nums[0],nums[0]]
        for i in range(len(nums)-1):
            if nums[i+1] - nums[i] == 1:
                temp[1] = nums[i+1]
            else:
                if temp[0] == temp[1]:
                    ret.append(str(temp[0]))
                else:
                    ret.append(str(temp[0]) + '->' + str(temp[1]))
                temp[0] = temp[1] = nums[i+1]
        if temp[0] == temp[1]:
            ret.append(str(temp[0]))
        else:
            ret.append(str(temp[0]) + '->' + str(temp[1]))
        return ret
obj = Solution()
print(obj.summaryRanges(nums = [0,1,2,4,5,7]))

['0->2', '4->5', '7']


In [ ]:
nums = [0,1,2,4,5,7]
ret = []
temp = [nums[0],0]
for i in range(len(nums)-1):
    if nums[i+1] - nums[i] == 1:
        temp[1] = nums[i]
    else:
        if temp[0] == temp[1]:
            ret.append(str(temp[0]))
        else:
            ret.append(str(temp[0]) + '->' + str(temp[1]))
        temp[0] = nums[i]
print(ret)


def summaryRanges(nums):
    ranges, r = [], []
    for n in nums:
        if n-1 not in r:
            r = []
            ranges += r,
        r[1:] = n,
    return ['->'.join(map(str, r)) for r in ranges]
summaryRanges(nums)

['0->1', '2->4']


['0->2', '4->5', '7']

# **Daily Coding Challenge March**

## ***Mar 3:* 413. Arithmetic Slices**

In [ ]:
class Solution:
    def numberOfArithmeticSlices(self, A) -> int:
        le=len(A)
        l=[0]*(le)
        for i in range(2,le):
            if A[i]-A[i-1] == A[i-1]-A[i-2]:
                l[i]=1+l[i-1]
        return sum(l)
obj = Solution()
print(obj.numberOfArithmeticSlices(A = [1,2,3,4]))

3


## ***Mar 4:* Problem statment**

## ***Mar 5:* Problem statment**

In [ ]:
class Solution:
    def deleteAndEarn(self, nums) -> int:
        """
            This question is similar to https://leetcode.com/problems/house-robber.
            The only difference here is we need to first calculate the money in each house.
            Let us consider, we have houses from 0 to max(nums). Now each time a number occurs in nums, we add that number to its corresponding house.
            At the end we will left with houses and the money they have.
            
        """
        lookup = [0] * (max(nums) + 1)
        for num in nums:
            lookup[num] += 1

        n = len(lookup)
        dp = [0] * n
        dp[0] = lookup[0]
        dp[1] = max(lookup[0], lookup[1])
        for i in range(2, n):
            dp[i] = max(i*lookup[i] + dp[i-2], dp[i-1])
        return dp[-1]
obj = Solution()
print(obj.deleteAndEarn(nums = [3,4,2]))

6


## ***Mar 6:* 1359. Count All Valid Pickup and Delivery Options**

In [ ]:
class Solution:
    def countOrders(self, n: int) -> int:
        a = 1
        for i in range(2, n+1):
            a *= i*(2*i-1)
        return a % (10**9 + 7)

obj = Solution()
print(obj.countOrders(800))

150560050


## ***Mar 7:* Problem statement**

## ***Mar 8:* 141. Linked List Cycle**

In [ ]:
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None
    def print_node(self):
        while self:
            # print(self.val, end = ' ')
            self = self.next

class Solution:
    def hasCycle(self, head) -> bool:
        slow = fast = head
        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next
            if slow == fast:
                return True
        return False
obj = Solution()
o1 = ListNode(3)
o2 = ListNode(2)
o3 = ListNode(0)
o4 = ListNode(-4)
o1.next = o2
o2.next = o3
o3.next = o4
o4.next = o2
head = o1
# head.print_node()
print(obj.hasCycle(head))

True


## ***Mar 13:* 20. Valid Parentheses**

In [ ]:
class Solution:
    def isValid(self, s: str) -> bool:
        if len(s) % 2 != 0:
            return False
        lists =[]  
        dic ={
            '}':'{',
            ')':'(',
            ']':'['
        }
        dic_values = dic.values()  # [ '{', '(',  '[' ]
        for i in s:  # 
            if i in dic_values:
                lists.append(i)
            else:
                if len(lists) > 0 and lists[-1] ==dic[i]:
                    lists.pop()
                else:
                    return False
        return len(lists) == 0
obj = Solution()
print(obj.isValid('(]'))

False


# **Biweekly Contest 72**

## **2176. Count Equal and Divisible Pairs in an Array**

In [ ]:
class Solution:
    def countPairs(self, nums, k: int) -> int:
        if list(set(nums)) == nums:
            return 0
        l = len(nums)
        count = 0
        for i in range(l):
            for j in range(i+1 ,l):
                if nums[i] == nums[j]:
                    if (i * j) % k == 0:
                        count += 1
                        print(i,j)
        return count

obj = Solution()
print(obj.countPairs(nums = [3,1,2,2,2,1,3], k = 2))

0 6
2 3
2 4
3 4
4


## **2177. Find Three Consecutive Integers That Sum to a Given Number**

In [ ]:
n = 100
val = n // 3
print([val-1 ,val , val+1], sum([val-1 ,val , val+1]))

[32, 33, 34] 99


## **2178. Maximum Split of Positive Even Integers**

In [ ]:
class Solution:
    def maximumEvenSplit(self, finalSum: int):
        if finalSum % 2 != 0:
            return []
        x = 2
        ret = []
        while finalSum >= x * 2 + 2:
            ret.append(x)
            finalSum -= x
            x += 2
            print(x, finalSum, x * 2 + 2)
            ret.append(finalSum)
        return ret
obj = Solution()
print(obj.maximumEvenSplit(finalSum = 28))

4 26 10
6 22 14
8 16 18
[2, 26, 4, 22, 6, 16]


# **Weekly Contest 281**

## **6012. Count Integers With Even Digit Sum**

In [ ]:
class Solution:
    def countEven(self, num: int) -> int:
        def is_even(n):
            val = str(n)
            ret = 0
            for i in val:
                ret += int(i)
            #     print(i, end = ' ')
            # print(ret, end = ' ')
            if ret % 2 == 0:
                return True
            else:
                return False
        
        count = 0
        for i in range(1, num+1):
            if is_even(i):
                print(i)
                count += 1
        return count
obj = Solution()
print(obj.countEven(30))

2
4
6
8
11
13
15
17
19
20
22
24
26
28
14


## **6013. Merge Nodes in Between Zeros**

In [ ]:
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
    def print_Node(self):
        lst = []
        while self:
          lst.append(self.val)
          self = self.next
        return lst
class Solution:
    def mergeNodes(self, head):
        curr = head
        ret = ListNode(0)
        ret1 = ret
        array = []
        curr = curr.next
        while curr:
            if curr.val != 0:
                array.append(curr.val)
                curr = curr.next
            elif curr.val == 0:
                cur = ListNode(sum(array))
                ret.next = cur
                ret = ret.next
                curr = curr.next
                array = []
        return ret1.next
o1 = ListNode()
o2 = ListNode()
o3 = ListNode()
o4 = ListNode()
o5 = ListNode()
o6 = ListNode()
o7 = ListNode()
o8 = ListNode()
o1.val = 0
o2.val = 1
o3.val = 0
o4.val = 3
o5.val = 0
o6.val = 2
o7.val = 2
o8.val = 0
head = o1
o1.next = o2
o2.next = o3
o3.next = o4
o4.next = o5
o5.next = o6
o6.next = o7
o7.next = o8
o8.next = None

obj = Solution()
head1 = obj.mergeNodes(head)
head1.print_Node()
# print(obj.mergeNodes(head1))


[1, 3, 4]

# **Weekly Contest 282**

## **Weekly Contest 282**

In [ ]:
class Solution:
    def prefixCount(self, words, pref: str) -> int:
        l = len(pref)
        print(words[1][:2])
        count = 0
        for i in range(len(words)):
            if words[i][:l] == pref:
                count += 1
        return count
obj = Solution()
print(obj.prefixCount(words = ["pay","attention","practice","attend"], pref = "at"))

at
2


## **6009. Minimum Number of Steps to Make Two Strings Anagram II**

In [ ]:
from collections import *
class Solution:
    def minSteps(self, s: str, t: str) -> int:
        return len(list((Counter(t) - Counter(s)).elements())) + len(list((Counter(s) - Counter(t)).elements()))
obj = Solution()
print(obj.minSteps(s = "leetcode", t = "coats"))

7


In [ ]:
print(set("leetcode").symmetric_difference(set("coats")))
print(set("night").difference(set("thing")))

{'s', 'e', 'a', 'l', 'd'}
set()


In [ ]:
from collections import Counter
test_list1 = [1, 3, 4, 5, 1, 3, 3]
test_list2 = [1, 3, 5]  
res = list((Counter(test_list1) - Counter(test_list2)).elements())
res1 = list((Counter(test_list2) - Counter(test_list1)).elements())
print("The list after performing the subtraction : " + str(res))
print("The list after performing the subtraction : " + str(res1))

The list after performing the subtraction : [1, 3, 3, 4]
The list after performing the subtraction : []


In [ ]:
from collections import Counter
test_list1 = "leetcode"
test_list2 = 'coats'  
res = list((Counter("leetcode") - Counter("coats")).elements())
res1 = list((Counter("coats") - Counter("leetcode")).elements())
print("The list after performing the subtraction : " + str(res))
print("The list after performing the subtraction : " + str(res1))

The list after performing the subtraction : ['l', 'e', 'e', 'e', 'd']
The list after performing the subtraction : ['a', 's']


## **6010. Minimum Time to Complete Trips**

In [ ]:
class Solution:
    def minimumTime(self, time, totalTrips: int) -> int:
        lo, hi = 1, totalTrips * min(time)
    
        def f(x):
            return sum(x // t for t in time) >= totalTrips
        
        while lo < hi:
            mid = (lo + hi) // 2
            if not f(mid): lo = mid + 1
            else: hi = mid
        return lo
obj = Solution()
print(obj.minimumTime(time = [1,2,3], totalTrips = 5))

3


# **Biweekly Contest 73**

## **6024. Most Frequent Number Following Key In an Array**

In [ ]:
dic = {9: 2, 100: 2, 200: 1}
print(max(dic, key = dic.get))


9


In [ ]:
class Solution:
    def mostFrequent(self, nums, key: int) -> int:
        n = 0
        r = dict()
        for i in range(len(nums)-2):
            if nums[i] == key:
                nex = nums[i+1]
                if nums[i] not in r:
                    r[nums[i]] = 1
                else:
                    r[nums[i]] += 1
                j = i + 2
                while True:
                    if nums[j] == nex:
                        r[nums[i]] += 1
                        j += 1
                    else:
                        break
        return max(r, key = r.get)
obj = Solution()
print(obj.mostFrequent(nums = [2,2,2,2,3], key = 2))

2


In [ ]:
from collections import Counter
class Solution:
    def mostFrequent(self, nums, key: int) -> int:
        n = len(nums)
        C = Counter()
        for i in range(n-1):
            if nums[i] == key:
                C[nums[i+1]] += 1
        C = list(C.items())
        print(C)
        C.sort(key=lambda x:-x[1])
        return C[0][0]
obj = Solution()
print(obj.mostFrequent(nums = [11,22,11,33,11,33], key = 11))

[(22, 1), (33, 2)]
33


In [ ]:
class Solution:
    def mostFrequent(self, nums, key: int) -> int:
        n = len(nums)
        C = Counter()
        for i in range(n-1):
            if nums[i] == key:
                C[nums[i+1]] += 1
        C = list(C.items())
        print(C)
        res = -1; cnt = -1
        for i in range(len(C)):
            if cnt < C[i][1]:
                cnt = C[i][1]
                res = C[i][0]
        return res
obj = Solution()
print(obj.mostFrequent(nums = [11,22,11,33,11,33], key = 11))


[(22, 1), (33, 2)]
33


# **Weekly Contest 283**

## **6016. Cells in a Range on an Excel Sheet**

In [ ]:
val = "ZABCDZ"
val = val[:-1] + chr((ord(val[-1])) + 1)
print(val)

ZABCD


In [ ]:
class Solution:
    def cellsInRange(self, s: str):
        st, en = s.split(':')
        st_al = ''
        st_num = 0
        en_al = ''
        en_num = 0
        ret = []
        for i in range(len(st)):
            if st[i].isdigit():
                st_num = int(st[i:])
                break
            else:
                st_al += st[i]
        for i in range(len(en)):
            if en[i].isdigit():
                en_num = int(en[i:])
                break
            else:
                en_al += en[i]
        def increment(val):
            if val[-1] == 'Z': return val[:-1] + 'A'
            val = val[:-1] + chr(ord(val[-1]) + 1)
            return val
        j = st_num
        while st_al != en_al or not(st_num == en_num):
            if j < en_num:
                ret.append(st_al + str(j))
                j += 1
            elif j == en_num:
                ret.append(st_al + str(j))
                if st_al != en_al:
                    j = st_num
                    st_al = increment(st_al)
                else:
                    st_num = en_num
            else:
                break
        if st == en:    ret.append(st)
        if ret[-1] != en: ret.append(en)
        return ret
obj = Solution()
print(obj.cellsInRange(s = "A1:F1"))

['A1', 'B1', 'C1', 'D1', 'E1', 'F1']


# **Biweekly Contest 74**

## **6020. Divide Array Into Equal Pairs**

In [3]:
from collections import Counter
class Solution:
    def divideArray(self, nums) -> bool:
        n = len(nums) // 2
        counts = Counter(nums)
        for i, j in counts.items():
            if j % 2 != 0:
                return False
        return True
obj = Solution()
print(obj.divideArray(nums = [3,2,3,2,2,2]))

True


In [5]:
text = "iekbksdsmuuzwxbpmcngsfkjvpzuknqguzvzik"
pattern = "mp"
class Solution:
    def maximumSubsequenceCount(self, text: str, pattern: str) -> int:
        new_text = ''
        for i in range(len(text)):
            if text[i] in pattern:
                new_text += text[i]
        counts = Counter(new_text)
        print(new_text)
        val1 = counts[pattern[0]]
        val2 = counts[pattern[1]]
        value = (min(val1, val2) + 1) * max(val1, val2)
        return value
obj = Solution()
print(obj.maximumSubsequenceCount(text, pattern))

mpmp
6


In [15]:
text = "iekbksdsmuuzwxbpmcngsfkjvpzuknqguzvzik"; pattern = "mp"
class Solution:
    def maximumSubsequenceCount(self, text: str, p: str) -> int:
        if p[0] == p[1]:
            s = text.count(p[0])+1
            return s*(s-1)//2
        n = len(text)
        a = b = 0
        for i in text:
            if i == p[0]:
                a += 1
            if i == p[1]:
                b += a
        print(a, b)
        return b+max(text.count(p[0]), text.count(p[1]))

obj = Solution()
print(obj.maximumSubsequenceCount(text, pattern))

2 3
5
